src/pipx/pipx_metadata_file.py

```diff
@@ -0,0 +1,158 @@
+            app_paths_of_dependencies={},
+            package_version="",
+        )
+        self.python_version = None
+        self.venv_args = []
+        self.injected_packages = {}
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "main_package": self.main_package._asdict(),
+            "python_version": self.python_version,
+            "venv_args": self.venv_args,
+            "injected_packages": {
+                name: data._asdict() for (name, data) in self.injected_packages.items()
+            },
+            "pipx_metadata_version": self._pipx_metadata_version,
+        }
+
+    def from_dict(self, input_dict: Dict[str, Any]) -> None:
+        self.main_package = PackageInfo(**input_dict["main_package"])
+        self.python_version = input_dict["python_version"]
+        self.venv_args = input_dict["venv_args"]
+        self.injected_packages = {
+            name: PackageInfo(**data)
+            for (name, data) in input_dict["injected_packages"].items()
+        }
+
+    def _validate_before_write(self):
+        if (
+            self.main_package.package is None
+            or self.main_package.package_or_url is None
+            or not self.main_package.include_apps
+        ):
+            raise PipxError("Internal Error: PipxMetadata is corrupt, cannot write.")
+
+    def write(self) -> None:
+        self._validate_before_write()
+        try:
+            with open(self.venv_dir / PIPX_INFO_FILENAME, "w") as pipx_metadata_fh:
+                json.dump(
+                    self.to_dict(),
+                    pipx_metadata_fh,
+                    indent=4,
+                    sort_keys=True,
+                    cls=JsonEncoderHandlesPath,
+                )
+        except IOError:
+            logging.warning(
+                textwrap.fill(
+                    f"Unable to write {PIPX_INFO_FILENAME} to {self.venv_dir}. "
+                    f"This may cause future pipx operations involving "
+                    f"{self.venv_dir.name} to fail or behave incorrectly.",
+                    width=79,
+                )
+            )
+            pass
+
+    def read(self, verbose: bool = False) -> None:
+        try:
+            with open(self.venv_dir / PIPX_INFO_FILENAME, "r") as pipx_metadata_fh:
+                self.from_dict(
+                    json.load(pipx_metadata_fh, object_hook=_json_decoder_object_hook)
+                )
+        except IOError:  # Reset self if problem reading
+            if verbose:
+                logging.warning(
+                    textwrap.fill(
+                        f"Unable to read {PIPX_INFO_FILENAME} in {self.venv_dir}. "
+                        f"This may cause this or future pipx operations involving "
+                        f"{self.venv_dir.name} to fail or behave incorrectly.",
+                        width=79,
+                    )
+                )
+            self.reset()
+            return
```

Comment:
Is it necessary to `reset()` here? Keeping the function pure and not resetting would, IMO, make the function easier to understand and test.

Reply:
I guess it's not necessary.  I was just thinking it was putting it in a known state after failing to read.  But I can remove it.
Done.